format/spec.md

```diff
@@ -1370,3 +1370,16 @@ Writing v2 metadata:
     * `sort_columns` was removed
 
 Note that these requirements apply when writing data to a v2 table. Tables that are upgraded from v1 may contain metadata that does not follow these requirements. Implementations should remain backward-compatible with v1 metadata requirements.
+
+## Appendix F: Implementation Notes
+
+This section covers topics not required by the specification but recommendations for systems implementing the Iceberg specification
+to help maintain a uniform experience.
+
+### Point in Time Reads (Time Travel)
+
+Iceberg supports two types of histories for tables. A history of previous "current snapshots" stored in ["snapshot-log" table metadata](#table-metadata-fields) and [parent-child lineage stored in "snapshots"](#table-metadata-fields). These two histories 
+might indicate different snapshot IDs for a specific timestamp. The discrepancies can be caused by a variety of table operations (e.g. updating the `current-snapshot-id` of the table).
+
+When processing point in time queries the Iceberg community has chosen to use "snapshot-log" metadata to lookup the table state
```

Comment:
nit: maybe `For point in time queries, "snapshot-log" metadata are used to lookup the table state at the given point of time`?

Reply:
I'm not sure we would say "the iceberg community has chosen." If we write it into the spec that's redundant. Also this isn't an implementation note though if we are telling folks to do this.

Reply:
I agree with Russell, but I would not block this clarification from going in. @RussellSpitzer, is this a blocker for you? If so could you note it on the vote thread? I'm going to vote +1, but we should wait until you're satisfied.

Reply:
I'll rephrase when I move it under snapshots.

Reply:
+1 for removing "the Iceberg community has chosen to", the spec should not involve the community, we can just describe what should be done.

Reply:
imho, spec should not contain "community has chosen": the spec should only focus on what's expected.

Reply:
@emkornfield can you address this before merging?

Reply:
Yes, I will do so.  I'd like to try to come to a resolution on https://github.com/apache/iceberg/pull/8982#discussion_r1678756567 so we can finalize in outstanding items in one commit.

Reply:
I voted +1 on the email thread. I would like this cleaned up though before we do the final merge.

Reply:
Since it is an optional field may be we should add "if exists" for "snapshot-log"?

Reply:
I think this recommendation affects REST catalog implementations.

I believe this text is not normative, therefore REST catalog implementations that do not expose the full snapshot history in metadata JSON will not violate the spec. 

However, for the sake of clarity and considering that the field is option, I propose adding a phrase like "when the catalog makes the snapshot history available in the metadata JSON, time travel queries should be executed like this.... [existing spec text]. If a catalog does not expose the full snapshot history, time travel queries should provide clear messages in case they cannot find the appropriate snapshot".

Reply:
IIUC conformant REST catalogs as of now must have a link to require the JSON metadata file to be present, so I don't think this directly reflects them.  If REST catalogs do not have a first class REST API for retrieving this information (or a way to parameterize the API via a timestamp to get the metadata) that seems like a new feature for the REST catalog, which is out of scope.  Hopefully the proposal to remove the requirement for JSON metadata  file is not approved until issues like this are sorted out.

Reply:
@emkornfield : my comment above is not about the retrieval of the JSON metadata itself (which must be supported), but about the (optional) `snapshot-log` field in the JSON payload returned by a REST Catalog server.

Reply:
This spec is independent of the REST catalog protocol. The protocol covers how to exchange table information covered by this spec, but this spec covers how to track that information and, in this case, recommendations for how that information is used. In the context of your suggestion, this assumes that "when the catalog makes the snapshot history available in the metadata JSON" is all the time. It is always true because this spec defines the metadata JSON.

I think it is valuable to say that engines are encouraged to use the information from `snapshot-log` for time travel by timestamp so that the results match what a query would have seen at that time. We made that choice for Spark because we think that is what users expect.

Reply:
`snapshot-log` is optional in both v1 and v2 spec: https://iceberg.apache.org/spec/#table-metadata-fields

Given that this PR suggests how engines should process that information so that users can expect similar behaviour from different engines, it would be nice to provide a suggestion of how to deal with missing `snapshot-log` in a common way too. In other words, how to produce a user-friendly message as opposed to some internal and engine-specific error.

Reply:
Added sentence mentioning an informative message being raised to the user about missing metadata.  I also tried to rephrased to remove the the iceberg community. If further word-smithing is desired, suggestions are welcome.

Reply:
Thanks for the update @emkornfield - current text LGTM :+1: